graveler/committed/merge_iterator.go

```diff
@@ -0,0 +1,109 @@
+package committed
+
+import (
+	"bytes"
+
+	"github.com/treeverse/lakefs/graveler"
+)
+
+type mergeIterator struct {
+	diffIt graveler.DiffIterator
+	val    *graveler.ValueRecord
+	base   Iterator
+	err    error
+}
+
+// NewMergeIterator accepts an iterator describing a diff from theirs to ours.
+// It returns a ValueIterator with the changes to perform on theirs, in order to merge ours into it,
+// relative to base as the merge base.
+// The iterator will return ErrConflictFound when it reaches a conflict.
+func NewMergeIterator(diffTheirsToOurs graveler.DiffIterator, base Iterator) (*mergeIterator, error) {
+	return &mergeIterator{diffIt: diffTheirsToOurs, base: base}, nil
+}
+
+func (d *mergeIterator) valueFromBase(key graveler.Key) *graveler.ValueRecord {
+	d.base.SeekGE(key)
+	var val *graveler.ValueRecord
+	for d.base.Next() && val == nil {
+		val, _ = d.base.Value()
+	}
+	if val == nil || !bytes.Equal(val.Key, key) {
+		return nil
+	}
+	return val
+}
+
+func (d *mergeIterator) Next() bool {
+	for d.diffIt.Next() {
+		val := d.diffIt.Value()
+		key := val.Key
+		typ := val.Type
+		baseVal := d.valueFromBase(key)
+		switch typ {
+		case graveler.DiffTypeAdded:
+			if baseVal == nil {
+				d.setValue()
+				return true
+			}
+			if !bytes.Equal(baseVal.Identity, val.Value.Identity) {
```

Comment:
if I read this correctly, this means we got to a record where their identity for a key is different from the base identity for a key. why is it a conflict?